src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -69,12 +87,63 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 	/* ensure that the task is not NULL */
 	Assert(task != NULL);
 
-	/* disallow transactions and triggers during distributed modify commands */
+	/* disallow triggers during distributed modify commands */
 	if (commandType != CMD_SELECT)
 	{
-		bool topLevel = true;
-		PreventTransactionChain(topLevel, "distributed commands");
 		eflags |= EXEC_FLAG_SKIP_TRIGGERS;
+
+		if (IsTransactionBlock() && xactParticipantHash == NULL)
+		{
+			MemoryContext oldContext = NULL;
+			ListCell *placementCell = NULL;
+
+			xactParticipantHash = CreateXactParticipantHash();
+
+			oldContext = MemoryContextSwitchTo(TopTransactionContext);
+
+			foreach(placementCell, task->taskPlacementList)
+			{
+				ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+				XactParticipantKey participantKey;
+				XactParticipantEntry *participantEntry = NULL;
+				bool entryFound = false;
+
+				PGconn *connection = NULL;
+				uint64 *shardIdPtr = NULL;
+
+				MemSet(&participantKey, 0, sizeof(participantKey));
+				strncpy(participantKey.nodeName, placement->nodeName, MAX_NODE_LENGTH);
```

Comment:
if you want to use a limited form of strcpy, it's generally better to use strlcpy, that guarantees to write a trailing 0. In this case the memset should guarantee the presence, but ...